### ISSUE TITLE
Update to Electron 22

### ISSUE THREAD

**Issue by deepak1556** [microsoft/vscode#166265]
- [x] Update patches
- [x] Nightly Exploration builds
- [x] Update native modules
- [x] Identify regressions from v8 sandbox
- [x] Update/Check LICENSE file in CG for new node.js version
- [x] Update LICENSE for new ffmpeg version
- [x] Upload symbols to appcenter

**Comment by mozhuanzuojing** [comment]
eletron22 released yesterday

**Comment by PF4Public** [comment]
@deepak1556 Have you attempted building `electron-22.x.y` branch? For me vscode lacks terminal. Could be `node-pty` or `vscode` issue?

**Comment by deepak1556** [comment]
Which OS are you seeing the issue on ? Did all native modules compile successfully during `yarn` step ?

**Comment by PF4Public** [comment]
> Which OS are you seeing the issue on ? Did all native modules compile successfully during `yarn` step ?

Linux. Yes, everything was compiled successfully but there was an error in Devtools console: something about posix spawn. I have rolled back to electron-21, but can re-test again at a later time.

**Comment by PF4Public** [comment]
Tried building vscode again, but it failed this time with `tree-sitter`. I remember it also failed last time, but skipped as it is an optional dependency. Now it fails further down the road:
```sh
[18:46:44] Starting 'vscode-linux-x64-min'...
[18:46:44] Starting clean-out-build ...
[18:46:44] Finished clean-out-build after 2 ms
[18:46:44] Starting build-web-node-paths ...
[18:46:45] Finished build-web-node-paths after 3 ms
[18:46:45] Starting compile-api-proposal-names ...
[18:46:45] Starting compilation api-proposal-names...
Assertion failed: bad export type for `tree_sitter_tsx_external_scanner_create`: undefined
Unhandled Rejection at: Promise Promise {
  <rejected> RuntimeError: abort(Assertion failed: bad export type for `tree_sitter_tsx_external_scanner_create`: undefined). Build with -s ASSERTIONS=1 for more info.
      at se (/var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/web-tree-sitter/tree-sitter.js:1:6761)
      at k (/var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/web-tree-sitter/tree-sitter.js:1:4462)
      at Pe (/var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/web-tree-sitter/tree-sitter.js:1:12582)
      at c (/var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/web-tree-sitter/tree-sitter.js:1:10613)
      at /var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/web-tree-sitter/tree-sitter.js:1:10879
      at async /var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/@vscode/l10n-dev/dist/main.js:254:10
      at async /var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/@vscode/l10n-dev/dist/main.js:243:22
} reason: RuntimeError: abort(Assertion failed: bad export type for `tree_sitter_tsx_external_scanner_create`: undefined). Build with -s ASSERTIONS=1 for more info.
    at se (/var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/web-tree-sitter/tree-sitter.js:1:6761)
    at k (/var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/web-tree-sitter/tree-sitter.js:1:4462)
    at Pe (/var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/web-tree-sitter/tree-sitter.js:1:12582)
    at c (/var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/web-tree-sitter/tree-sitter.js:1:10613)
    at /var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/web-tree-sitter/tree-sitter.js:1:10879
    at async /var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/@vscode/l10n-dev/dist/main.js:254:10
    at async /var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/@vscode/l10n-dev/dist/main.js:243:22
[18:46:45] The following tasks did not complete: vscode-linux-x64-min
[18:46:45] Did you forget to signal async completion?
```
The issue with tree-sitter is as follows:
```sh
warning Error running install script for optional dependency: /var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/build/node_modules/tree-sitter: Command failed.
Exit code: 1
Command: prebuild-install || node-gyp rebuild
Arguments: 
Directory: /var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/build/node_modules/tree-sitter
Output:
gyp info it worked if it ends with ok
gyp info using node-gyp@9.0.0
gyp info using node@16.17.1 | linux | x64
gyp info find Python using Python version 3.10.9 found at \"/usr/bin/python3.10\"
gyp WARN read config.gypi ENOENT: no such file or directory, open '/usr/include/electron-22/node/include/node/config.gypi'
gyp info spawn /usr/bin/python3.10
gyp info spawn args [
gyp info spawn args   '/usr/lib64/electron-22/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '/var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/build/node_modules/tree-sitter/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/usr/lib64/electron-22/node_modules/npm/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/usr/include/electron-22/node/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=/usr/include/electron-22/node',
gyp info spawn args   '-Dnode_gyp_dir=/usr/lib64/electron-22/node_modules/npm/node_modules/node-gyp',
gyp info spawn args   '-Dnode_lib_file=/usr/include/electron-22/node/$(Configuration)/node.lib',
gyp info spawn args   '-Dmodule_root_dir=/var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/build/node_modules/tree-sitter',
gyp info spawn args   '-Dnode_engine=v8',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.'
gyp info spawn args ]
gyp info spawn make
gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
make: Entering directory '/var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/build/node_modules/tree-sitter/build'
  CC(target) Release/obj.target/tree_sitter/vendor/tree-sitter/lib/src/lib.o
  AR(target) Release/obj.target/tree_sitter.a
  COPY Release/tree_sitter.a
  CXX(target) Release/obj.target/tree_sitter_runtime_binding/src/binding.o
In file included from /usr/include/electron-22/node/v8-object.h:10,
                 from /usr/include/electron-22/node/v8-array-buffer.h:13,
                 from /usr/include/electron-22/node/v8.h:24,
                 from /usr/include/electron-22/node/node.h:79,
                 from ../src/binding.cc:1:
/usr/include/electron-22/node/v8-persistent-handle.h:256:44: warning: ‘template<class T> struct v8::CopyablePersistentTraits’ is deprecated: Use v8::Global instead [-Wdeprecated-declarations]
  256 |   using CopyablePersistent = Persistent<T, CopyablePersistentTraits<T>>;
      |                                            ^~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/electron-22/node/v8-persistent-handle.h:255:48: note: declared here
  255 | struct V8_DEPRECATED(\"Use v8::Global instead\") CopyablePersistentTraits {
      |                                                ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../../nan/nan.h:178,
                 from ../src/./language.h:4,
                 from ../src/binding.cc:3:
../../nan/nan_callbacks.h:55:23: error: ‘AccessorSignature’ is not a member of ‘v8’
   55 | typedef v8::Local<v8::AccessorSignature> Sig;
      |                       ^~~~~~~~~~~~~~~~~
../../nan/nan_callbacks.h:55:40: error: template argument 1 is invalid
   55 | typedef v8::Local<v8::AccessorSignature> Sig;
      |                                        ^
In file included from ../src/./language.h:4,
                 from ../src/binding.cc:3:
../../nan/nan.h:204:16: warning: ‘template<class T> struct v8::CopyablePersistentTraits’ is deprecated: Use v8::Global instead [-Wdeprecated-declarations]
  204 |     public v8::CopyablePersistentTraits<T> {};
      |                ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/electron-22/node/v8-object.h:10,
                 from /usr/include/electron-22/node/v8-array-buffer.h:13,
                 from /usr/include/electron-22/node/v8.h:24,
                 from /usr/include/electron-22/node/node.h:79,
                 from ../src/binding.cc:1:
/usr/include/electron-22/node/v8-persistent-handle.h:255:48: note: declared here
  255 | struct V8_DEPRECATED(\"Use v8::Global instead\") CopyablePersistentTraits {
      |                                                ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../src/./language.h:4,
                 from ../src/binding.cc:3:
../../nan/nan.h: In function ‘void Nan::SetAccessor(v8::Local<v8::ObjectTemplate>, v8::Local<v8::String>, Nan::GetterCallback, Nan::SetterCallback, v8::Local<v8::Value>, v8::AccessControl, v8::PropertyAttribute, Nan::imp::Sig)’:
../../nan/nan.h:2544:19: error: no matching function for call to ‘v8::ObjectTemplate::SetAccessor(v8::Local<v8::String>&, void (*&)(v8::Local<v8::Name>, const v8::PropertyCallbackInfo<v8::Value>&), void (*&)(v8::Local<v8::Name>, v8::Local<v8::Value>, const v8::PropertyCallbackInfo<void>&), v8::Local<v8::Object>&, v8::AccessControl&, v8::PropertyAttribute&, Nan::imp::Sig&)’
 2544 |   tpl->SetAccessor(
      |   ~~~~~~~~~~~~~~~~^
 2545 |       name
      |       ~~~~         
 2546 |     , getter_
      |     ~~~~~~~~~      
 2547 |     , setter_
      |     ~~~~~~~~~      
 2548 |     , obj
      |     ~~~~~          
 2549 |     , settings
      |     ~~~~~~~~~~     
 2550 |     , attribute
      |     ~~~~~~~~~~~    
 2551 |     , signature);
      |     ~~~~~~~~~~~~   
In file included from /usr/include/electron-22/node/v8-function.h:15,
                 from /usr/include/electron-22/node/v8.h:33,
                 from /usr/include/electron-22/node/node.h:79,
                 from ../src/binding.cc:1:
/usr/include/electron-22/node/v8-template.h:807:8: note: candidate: ‘void v8::ObjectTemplate::SetAccessor(v8::Local<v8::String>, v8::AccessorGetterCallback, v8::AccessorSetterCallback, v8::Local<v8::Value>, v8::AccessControl, v8::PropertyAttribute, v8::SideEffectType, v8::SideEffectType)’
  807 |   void SetAccessor(
      |        ^~~~~~~~~~~
/usr/include/electron-22/node/v8-template.h:812:22: note:   no known conversion for argument 7 from ‘Nan::imp::Sig’ {aka ‘int’} to ‘v8::SideEffectType’
  812 |       SideEffectType getter_side_effect_type = SideEffectType::kHasSideEffect,
      |       ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/electron-22/node/v8-template.h:814:8: note: candidate: ‘void v8::ObjectTemplate::SetAccessor(v8::Local<v8::Name>, v8::AccessorNameGetterCallback, v8::AccessorNameSetterCallback, v8::Local<v8::Value>, v8::AccessControl, v8::PropertyAttribute, v8::SideEffectType, v8::SideEffectType)’
  814 |   void SetAccessor(
      |        ^~~~~~~~~~~
/usr/include/electron-22/node/v8-template.h:819:22: note:   no known conversion for argument 7 from ‘Nan::imp::Sig’ {aka ‘int’} to ‘v8::SideEffectType’
  819 |       SideEffectType getter_side_effect_type = SideEffectType::kHasSideEffect,
      |       ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../src/binding.cc:1:
../src/binding.cc: At global scope:
/usr/include/electron-22/node/node.h:874:7: warning: cast between incompatible function types from ‘void (*)(v8::Local<v8::Object>)’ to ‘node::addon_register_func’ {aka ‘void (*)(v8::Local<v8::Object>, v8::Local<v8::Value>, void*)’} [-Wcast-function-type]
  874 |       (node::addon_register_func) (regfunc),                          \\\n      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/electron-22/node/node.h:908:3: note: in expansion of macro ‘NODE_MODULE_X’
  908 |   NODE_MODULE_X(modname, regfunc, NULL, 0)  // NOLINT (readability/null_usage)
      |   ^~~~~~~~~~~~~
../src/binding.cc:25:1: note: in expansion of macro ‘NODE_MODULE’
   25 | NODE_MODULE(tree_sitter_runtime_binding, InitAll)
      | ^~~~~~~~~~~
make: *** [tree_sitter_runtime_binding.target.mk:139: Release/obj.target/tree_sitter_runtime_binding/src/binding.o] Error 1
make: Leaving directory '/var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/build/node_modules/tree-sitter/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/lib64/electron-22/node_modules/npm/node_modules/node-gyp/lib/build.js:194:23)
gyp ERR! stack     at ChildProcess.emit (node:events:513:28)
gyp ERR! stack     at ChildProcess._handle.onexit (node:internal/child_process:291:12)
gyp ERR! System Linux 5.15.80-gentoo
gyp ERR! command \"/usr/lib64/electron-22/electron\" \"/usr/lib64/electron-22/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js\" \"rebuild\"
gyp ERR! cwd /var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/build/node_modules/tree-sitter
gyp ERR! node -v v16.17.1



warning Error running install script for optional dependency: /var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/build/node_modules/tree-sitter-typescript: Command failed.
Exit code: 1
Command: node-gyp rebuild
Arguments: 
Directory: /var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/build/node_modules/tree-sitter-typescript
Output:
gyp info it worked if it ends with ok
gyp info using node-gyp@9.0.0
gyp info using node@16.17.1 | linux | x64
gyp info find Python using Python version 3.10.9 found at \"/usr/bin/python3.10\"
gyp WARN read config.gypi ENOENT: no such file or directory, open '/usr/include/electron-22/node/include/node/config.gypi'
gyp info spawn /usr/bin/python3.10
gyp info spawn args [
gyp info spawn args   '/usr/lib64/electron-22/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '/var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/build/node_modules/tree-sitter-typescript/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/usr/lib64/electron-22/node_modules/npm/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/usr/include/electron-22/node/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=/usr/include/electron-22/node',
gyp info spawn args   '-Dnode_gyp_dir=/usr/lib64/electron-22/node_modules/npm/node_modules/node-gyp',
gyp info spawn args   '-Dnode_lib_file=/usr/include/electron-22/node/$(Configuration)/node.lib',
gyp info spawn args   '-Dmodule_root_dir=/var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/build/node_modules/tree-sitter-typescript',
gyp info spawn args   '-Dnode_engine=v8',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.'
gyp info spawn args ]
gyp info spawn make
gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
make: Entering directory '/var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/build/node_modules/tree-sitter-typescript/build'
  CC(target) Release/obj.target/tree_sitter_typescript_binding/typescript/src/parser.o
In file included from ../typescript/src/parser.c:1:
../typescript/src/parser.c: In function ‘ts_lex_keywords’:
../typescript/src/tree_sitter/parser.h:135:8: warning: variable ‘eof’ set but not used [-Wunused-but-set-variable]
  135 |   bool eof = false;             \\\n      |        ^~~
../typescript/src/parser.c:5342:3: note: in expansion of macro ‘START_LEXER’
 5342 |   START_LEXER();
      |   ^~~~~~~~~~~
  CC(target) Release/obj.target/tree_sitter_typescript_binding/typescript/src/scanner.o
  CC(target) Release/obj.target/tree_sitter_typescript_binding/tsx/src/parser.o
In file included from ../tsx/src/parser.c:1:
../tsx/src/parser.c: In function ‘ts_lex_keywords’:
../typescript/src/tree_sitter/parser.h:135:8: warning: variable ‘eof’ set but not used [-Wunused-but-set-variable]
  135 |   bool eof = false;             \\\n      |        ^~~
../tsx/src/parser.c:5600:3: note: in expansion of macro ‘START_LEXER’
 5600 |   START_LEXER();
      |   ^~~~~~~~~~~
  CC(target) Release/obj.target/tree_sitter_typescript_binding/tsx/src/scanner.o
  CXX(target) Release/obj.target/tree_sitter_typescript_binding/bindings/node/binding.o
In file included from /usr/include/electron-22/node/v8-object.h:10,
                 from /usr/include/electron-22/node/v8-array-buffer.h:13,
                 from /usr/include/electron-22/node/v8.h:24,
                 from /usr/include/electron-22/node/node.h:79,
                 from ../bindings/node/binding.cc:2:
/usr/include/electron-22/node/v8-persistent-handle.h:256:44: warning: ‘template<class T> struct v8::CopyablePersistentTraits’ is deprecated: Use v8::Global instead [-Wdeprecated-declarations]
  256 |   using CopyablePersistent = Persistent<T, CopyablePersistentTraits<T>>;
      |                                            ^~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/electron-22/node/v8-persistent-handle.h:255:48: note: declared here
  255 | struct V8_DEPRECATED(\"Use v8::Global instead\") CopyablePersistentTraits {
      |                                                ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../../nan/nan.h:178,
                 from ../bindings/node/binding.cc:3:
../../nan/nan_callbacks.h:55:23: error: ‘AccessorSignature’ is not a member of ‘v8’
   55 | typedef v8::Local<v8::AccessorSignature> Sig;
      |                       ^~~~~~~~~~~~~~~~~
../../nan/nan_callbacks.h:55:40: error: template argument 1 is invalid
   55 | typedef v8::Local<v8::AccessorSignature> Sig;
      |                                        ^
In file included from ../bindings/node/binding.cc:3:
../../nan/nan.h:204:16: warning: ‘template<class T> struct v8::CopyablePersistentTraits’ is deprecated: Use v8::Global instead [-Wdeprecated-declarations]
  204 |     public v8::CopyablePersistentTraits<T> {};
      |                ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/electron-22/node/v8-object.h:10,
                 from /usr/include/electron-22/node/v8-array-buffer.h:13,
                 from /usr/include/electron-22/node/v8.h:24,
                 from /usr/include/electron-22/node/node.h:79,
                 from ../bindings/node/binding.cc:2:
/usr/include/electron-22/node/v8-persistent-handle.h:255:48: note: declared here
  255 | struct V8_DEPRECATED(\"Use v8::Global instead\") CopyablePersistentTraits {
      |                                                ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../bindings/node/binding.cc:3:
../../nan/nan.h: In function ‘void Nan::SetAccessor(v8::Local<v8::ObjectTemplate>, v8::Local<v8::String>, Nan::GetterCallback, Nan::SetterCallback, v8::Local<v8::Value>, v8::AccessControl, v8::PropertyAttribute, Nan::imp::Sig)’:
../../nan/nan.h:2544:19: error: no matching function for call to ‘v8::ObjectTemplate::SetAccessor(v8::Local<v8::String>&, void (*&)(v8::Local<v8::Name>, const v8::PropertyCallbackInfo<v8::Value>&), void (*&)(v8::Local<v8::Name>, v8::Local<v8::Value>, const v8::PropertyCallbackInfo<void>&), v8::Local<v8::Object>&, v8::AccessControl&, v8::PropertyAttribute&, Nan::imp::Sig&)’
 2544 |   tpl->SetAccessor(
      |   ~~~~~~~~~~~~~~~~^
 2545 |       name
      |       ~~~~         
 2546 |     , getter_
      |     ~~~~~~~~~      
 2547 |     , setter_
      |     ~~~~~~~~~      
 2548 |     , obj
      |     ~~~~~          
 2549 |     , settings
      |     ~~~~~~~~~~     
 2550 |     , attribute
      |     ~~~~~~~~~~~    
 2551 |     , signature);
      |     ~~~~~~~~~~~~   
In file included from /usr/include/electron-22/node/v8-function.h:15,
                 from /usr/include/electron-22/node/v8.h:33,
                 from /usr/include/electron-22/node/node.h:79,
                 from ../bindings/node/binding.cc:2:
/usr/include/electron-22/node/v8-template.h:807:8: note: candidate: ‘void v8::ObjectTemplate::SetAccessor(v8::Local<v8::String>, v8::AccessorGetterCallback, v8::AccessorSetterCallback, v8::Local<v8::Value>, v8::AccessControl, v8::PropertyAttribute, v8::SideEffectType, v8::SideEffectType)’
  807 |   void SetAccessor(
      |        ^~~~~~~~~~~
/usr/include/electron-22/node/v8-template.h:812:22: note:   no known conversion for argument 7 from ‘Nan::imp::Sig’ {aka ‘int’} to ‘v8::SideEffectType’
  812 |       SideEffectType getter_side_effect_type = SideEffectType::kHasSideEffect,
      |       ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/electron-22/node/v8-template.h:814:8: note: candidate: ‘void v8::ObjectTemplate::SetAccessor(v8::Local<v8::Name>, v8::AccessorNameGetterCallback, v8::AccessorNameSetterCallback, v8::Local<v8::Value>, v8::AccessControl, v8::PropertyAttribute, v8::SideEffectType, v8::SideEffectType)’
  814 |   void SetAccessor(
      |        ^~~~~~~~~~~
/usr/include/electron-22/node/v8-template.h:819:22: note:   no known conversion for argument 7 from ‘Nan::imp::Sig’ {aka ‘int’} to ‘v8::SideEffectType’
  819 |       SideEffectType getter_side_effect_type = SideEffectType::kHasSideEffect,
      |       ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../bindings/node/binding.cc:2:
../bindings/node/binding.cc: At global scope:
/usr/include/electron-22/node/node.h:874:7: warning: cast between incompatible function types from ‘void (*)(v8::Local<v8::Object>, v8::Local<v8::Object>)’ to ‘node::addon_register_func’ {aka ‘void (*)(v8::Local<v8::Object>, v8::Local<v8::Value>, void*)’} [-Wcast-function-type]
  874 |       (node::addon_register_func) (regfunc),                          \\\n      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/electron-22/node/node.h:908:3: note: in expansion of macro ‘NODE_MODULE_X’
  908 |   NODE_MODULE_X(modname, regfunc, NULL, 0)  // NOLINT (readability/null_usage)
      |   ^~~~~~~~~~~~~
../bindings/node/binding.cc:35:1: note: in expansion of macro ‘NODE_MODULE’
   35 | NODE_MODULE(tree_sitter_typescript_binding, Init)
      | ^~~~~~~~~~~
make: *** [tree_sitter_typescript_binding.target.mk:129: Release/obj.target/tree_sitter_typescript_binding/bindings/node/binding.o] Error 1
make: Leaving directory '/var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/build/node_modules/tree-sitter-typescript/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/lib64/electron-22/node_modules/npm/node_modules/node-gyp/lib/build.js:194:23)
gyp ERR! stack     at ChildProcess.emit (node:events:513:28)
gyp ERR! stack     at ChildProcess._handle.onexit (node:internal/child_process:291:12)
gyp ERR! System Linux 5.15.80-gentoo
gyp ERR! command \"/usr/lib64/electron-22/electron\" \"/usr/lib64/electron-22/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js\" \"rebuild\"
gyp ERR! cwd /var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/build/node_modules/tree-sitter-typescript
gyp ERR! node -v v16.17.1
Done in 16.50s.
```

Looks like it also needs to bump nan to avoid AccessorSignatures

**Comment by deepak1556** [comment]
Ah okay I didn't realize you are doing a full build rather than running out of sources, yeah the `tree-sitter` package needs a nan update.

**Comment by tjx666** [comment]
@deepak1556 Plan to support node18?

**Comment by deepak1556** [comment]
It will happen when we update to Electron 23, not before that :)

**Comment by PF4Public** [comment]
Adding `"nan": "^2.17.0"` to resolutions helps building tree-sitter, but this issue persists:
```sh
[16:14:25] Using gulpfile /var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/gulpfile.js
[16:14:25] Starting 'vscode-linux-x64-min'...
[16:14:25] Starting clean-out-build ...
[16:14:25] Finished clean-out-build after 6 ms
[16:14:25] Starting build-web-node-paths ...
[16:14:25] Finished build-web-node-paths after 5 ms
[16:14:25] Starting compile-api-proposal-names ...
[16:14:25] Starting compilation api-proposal-names...
Assertion failed: bad export type for `tree_sitter_tsx_external_scanner_create`: undefined
Unhandled Rejection at: Promise Promise {
  <rejected> RuntimeError: abort(Assertion failed: bad export type for `tree_sitter_tsx_external_scanner_create`: undefined). Build with -s ASSERTIONS=1 for more info.
```

**Comment by ReazerDev** [comment]
Everything works fine for me on arch 6.1.5.
Is there anything I can test to help get this PR ready for this month's release?

Electron 22 fixes the blurriness I experience on Wayland. It'd be amazing if this get's released this month! :)

**Comment by lorand-horvath** [comment]
> Electron 22 fixes the blurriness I experience on Wayland. It'd be amazing if this get's released this month! :)

Electron 22 will not be integrated in the January release, the upgrade is only being investigated (at least that's the plan for January), see Electron heading in https://github.com/microsoft/vscode/issues/170413

**Comment by deepak1556** [comment]
@ReazerDev the plan is to provide nightly exploration builds in January milestone which will then be tested during our iteration endgame. We will push the update to Insiders in February milestone depending on the results from testing. Currently the exploration builds were blocked on https://github.com/microsoft/vscode/pull/166430#issuecomment-1331203694 which was recently addressed in `22.0.2`, we can move forward on those builds this week :)

**Comment by lorand-horvath** [comment]
@deepak1556 Are nightly exploration builds available to the public? Would test such an Electron 22 build, if that would be of help.

**Comment by deepak1556** [comment]
We haven't published the exploration builds on the website but I will post the build links in this issue thread once they are available. Thanks for offering to test, it helps!

**Comment by lorand-horvath** [comment]
@deepak1556 Can you please provide the Windows x64 .zip version?

**Comment by deepak1556** [comment]
[win32-x64-archive](https://az764295.vo.msecnd.net/exploration/68c9758ab2187969747dbcdd18c24bb15a0b42c7/VSCode-win32-x64-1.75.0-exploration.zip)

**Comment by lorand-horvath** [comment]
Thanks @deepak1556, it's working, but it seems there are some issues, which I'll post here as I find them, if that's OK.

The select dropdown has some CSS-related issues, I assume:
![image](https://user-images.githubusercontent.com/72015221/213643300-afdd4d1d-e952-4083-afc9-8b01d8d82f00.png)

**Comment by deepak1556** [comment]
If possible, please open separate issues since we have `electron-22-update` label, it allows for grouping known issues from a particular Electron update. For the above issue, it should be addressed with next Electron 22 patch release via https://github.com/electron/electron/commit/55c818d0a84400621c0817c8a2e4cc17d9eb24c8

**Comment by lorand-horvath** [comment]
Just checked in latest Insiders and it shows the same issue as above (I think dropdown background is transparent). So this is also occurring on electron 19 and AFAICT it is a different issue from https://github.com/electron/electron/pull/36934
I will open a separate issue for this.

Other than that, the Electron 22 build is working very well (just updated to latest exploration build from today). It may be placebo, but it seems snappier than Electron 19 based Insiders version. Still testing... Thank you!

**Comment by bpasero** [comment]
@lorand-horvath please do, the issue is a regression from work we did independent from electron update.

**Comment by lorand-horvath** [comment]
@bpasero @deepak1556 CSS issue unrelated to Electron reported in https://github.com/microsoft/vscode/issues/171804

**Comment by PF4Public** [comment]
Just a FYI I still get `bad export type for tree_sitter_tsx_external_scanner_create: undefined` if building against electron-22.

**Comment by lorand-horvath** [comment]
@deepak1556 The last exploration build from Feb 3rd (Electron 22.2.0) is working well for me, but unfortunately it still has this bug: https://github.com/microsoft/vscode/issues/173455 & https://github.com/microsoft/vscode/issues/173548

Would it be possible to rebuild it / bringing it up to date, please? The latest Insiders from today (Feb 7th) has fixed the issue: https://github.com/microsoft/vscode/issues/173455#issuecomment-1420351966

**Comment by tjx666** [comment]
Electron 23 may need to wait after 3~4 month, include this month for 22, 1 or 2 months for explore 23, 1 month to update.

Another big change in electron 23 is node 18. I think this month explore electron 23 build, next month update 23 would be better. Anthing big improvement in electron 22?

**Comment by deepak1556** [comment]
We will track Electron 23 with exploration builds once Electron 22 is merged to Insiders. Currently, we are targeting Electron 22 for VSCode v1.77 , refs https://code.visualstudio.com/docs/supporting/faq#_can-i-run-vs-code-on-old-macos-versions

>  Anthing big improvement in electron 22?

From initial perf numbers we saw an improvement in the render process creation phase but at the same time we also saw a dip in the script parsing phase (currently investigating this bit). We also get the [utility process API](https://github.com/electron/electron/blob/main/docs/api/utility-process.md) by default which we have been patching in Electron 19.

**Comment by tjx666** [comment]
why not electron24 after 22 is merged into vscode, ref: https://www.electronjs.org/docs/latest/tutorial/electron-timelines

**Comment by lorand-horvath** [comment]
@tjx666 Hold your horses... Electron 24 is not even in beta, so anything can change until March the 7th. Also the bundled Node.js version is subject to change/TBD.

As you can see above, even Electron 22 integration has been moved from 'In Progress' to 'Blocked'...

**Comment by tjx666** [comment]
@lorand-horvath Electron22 will be merged to vscode in March and published in April of VSCode 1.77. The next version of electron to explore will start in April, and electron 24 stable will be publish in April. And the next version of electron will finally shipped in June/July which need about 3/4 mounth.

**Comment by lorand-horvath** [comment]
@tjx666 I know all of that since I can look at the timeline table and judge all kinds of values and dates and versions and whatnot. However, what I meant is that jumping on bleeding edge, so-called "stable" software, is not always the best choice, as any seasoned developer knows very well, so that's why I very much doubt that the VSCode engineers will update to Electron 24 this year.

**Comment by PF4Public** [comment]
Is there any solution for [`tree_sitter_tsx_external_scanner_create` compilation issue](https://github.com/microsoft/vscode/issues/166265#issuecomment-1362829787)?

**Comment by lorand-horvath** [comment]
@deepak1556 Fun fact: latest Exploration build from today has Electron version 22.10.0, whereas latest available is 22.2.1 - I assume it's a custom build, similar to Electron 19.1.10 from Insiders?

**Comment by deepak1556** [comment]
@lorand-horvath yup we are testing some PGO related numbers, it will be reverted to official version once the testing is complete.

**Comment by darkbasic** [comment]
Any news about the [tree_sitter_tsx_external_scanner_create  compilation issue](https://github.com/microsoft/vscode/issues/166265#issuecomment-1362829787)?

**Comment by PF4Public** [comment]
`vscode/l10n-dev-0.0.21` depends on `web-tree-sitter` which depends on `tree-sitter`, which hasn't been updated since ages and predictably does not build. Even though `master` branch from `tree-sitter/node-tree-sitter` does solve one issue, there is still another build failure which prevents it from building. 

I would speculate it all started with https://github.com/microsoft/vscode/commit/0510da84dd109e682a654465ff1bdc376b76e231 and indeed https://github.com/microsoft/vscode/commit/88c8146de77e3632c3f909e5f5d7ce8fc4fc4c74 is the last functioning commit for me.

**Comment by lorand-horvath** [comment]
@deepak1556 I see you had to revert to Electron 19 https://github.com/microsoft/vscode/commit/3a11b47eee0cd39d658e67e4afb0007b708a12b1 Are there some kind of issues with Electron 22?

**Comment by PF4Public** [comment]
> @deepak1556 I see you had to revert to Electron 19 [3a11b47](https://github.com/microsoft/vscode/commit/3a11b47eee0cd39d658e67e4afb0007b708a12b1) Are there some kind of issues with Electron 22?

At least one is just above your message ;)

**Comment by ollie-iterators** [comment]
Is one of the issues with this: https://github.com/microsoft/vscode/issues/171508?

**Comment by deepak1556** [comment]
Sorry for the silence on the update, we have been investigating the impact of V8 sandbox for our extensions over the last two weeks, you can find info on that at https://github.com/microsoft/vscode/issues/177338. We have plan to move the update forward, but we will not merge the update for March 2023 milestone.

**Comment by PF4Public** [comment]
In case anyone is interested, the error `tree_sitter_tsx_external_scanner_create` comes from https://github.com/microsoft/vscode-l10n/blob/b941cad5b485dad98dc33bfa43b2c0542f8f7e44/l10n-dev/src/ast/analyzer.ts#L41-L52
```js
	static #tsGrammar: Promise<Parser.Language> = (async () => {
		await initParser;
		return await Parser.Language.load(
			path.resolve(__dirname, 'tree-sitter-typescript.wasm')
		);
	})();
	static #tsxGrammar: Promise<Parser.Language> = (async () => {
		await initParser;
		return await Parser.Language.load(
			path.resolve(__dirname, 'tree-sitter-tsx.wasm')
		);
	})();
```
Both wasm files come pre-built from [@vscode/l10n-dev](https://www.npmjs.com/package/@vscode/l10n-dev?activeTab=code).
Changing the dependency to version 0.0.24 changes the error (although that may be a race condition between the two): 
```sh
Unhandled Rejection at: Promise Promise {
  <rejected> Error: bad export type for `tree_sitter_typescript_external_scanner_create`: undefined
      at reportUndefinedSymbols (/var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/web-tree-sitter/tree-sitter.js:1:19748)
      at postInstantiation (/var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/web-tree-sitter/tree-sitter.js:1:17027)
      at /var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/web-tree-sitter/tree-sitter.js:1:17752
      at async /var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/@vscode/l10n-dev/dist/main.js:270:10
      at async /var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/@vscode/l10n-dev/dist/main.js:259:22
} reason: Error: bad export type for `tree_sitter_typescript_external_scanner_create`: undefined
    at reportUndefinedSymbols (/var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/web-tree-sitter/tree-sitter.js:1:19748)
    at postInstantiation (/var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/web-tree-sitter/tree-sitter.js:1:17027)
    at /var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/web-tree-sitter/tree-sitter.js:1:17752
    at async /var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/@vscode/l10n-dev/dist/main.js:270:10
    at async /var/tmp/portage/app-editors/vscode-9999/work/vscode-9999/node_modules/@vscode/l10n-dev/dist/main.js:259:22
```
Lines 259 and 270 from the error look as follows:
```js
__privateAdd(ScriptAnalyzer, _tsxParser, (async () => {
  await initParser;
  const parser = new import_web_tree_sitter.default();
  parser.setLanguage(await __privateGet(_ScriptAnalyzer, _tsxGrammar));
  return parser;
})());

__privateAdd(ScriptAnalyzer, _tsxGrammar, (async () => {
  await initParser;
  return await import_web_tree_sitter.default.Language.load(
    path.resolve(__dirname, "tree-sitter-tsx.wasm")
  );
})());
```
In this case `tree_sitter_typescript_external_scanner_create` is absent from `tree-sitter-tsx.wasm`, but present in `tree-sitter-typescript.wasm`.

I have no idea, why is it trying to load a function that is absent from wasm file. Broken wasm files in  [@vscode/l10n-dev](https://www.npmjs.com/package/@vscode/l10n-dev?activeTab=code)?

**Comment by deepak1556** [comment]
Closing as we will be shipping E22 update with `v1.78`

**Comment by deepak1556** [comment]
@The-BDD-Coach it needs to be addressed in https://github.com/tree-sitter/tree-sitter-typescript/issues/244, you can add `allowUndefined: true` to the load arguments in https://github.com/tree-sitter/tree-sitter/blob/91e4d940169a0c0b024560632ef53c4f119117ca/lib/binding_web/binding.js#L911 to suppress the error temporarily

**Comment by nethershaw** [comment]
@lorand-horvath @deepak1556 I have personally wasted a _week_ of development time trying to repair a broken installation and build VSCode 1.78+ against Electron 22, 23, and 24 only to find that you ignored several warnings from @PF4Public and merged before https://github.com/tree-sitter/tree-sitter-typescript/issues/244 was resolved or even acknowledged where reported.

They commented several times and you didn't pay them any attention whatsoever.

You have my thanks as I task my build system with reverting to VSCode 1.77.3 and Electron 19 and I wait.

**Comment by deepak1556** [comment]
@nethershaw our own build systems did not encounter the issue and hence action was not taken in this issue, if you are building VSCode in a custom way that is not referred in the docs then there is not much we can do to help. Please also check https://github.com/microsoft/vscode-l10n/issues/104 where the issue was actually followed up on by @TylerLeonhardt .

**Comment by darkbasic** [comment]
This is also most likely related: https://github.com/microsoft/vscode/issues/182743
If you have any suggestions on how to get any kind of logs would be greatly appreciated because while electron-19 doesn't crashes vscode itself returns without showing anything in the logs.

> our own build systems did not encounter the issue and hence action was not taken

If I had read this out of context I would have said it smells like a big tech company who doesn't know anything about open source and completely ignores the community, distribution maintainers, etc.  While it definitely doesn't apply to vscode it's still sad seeing such bug reports being so easily dismissed.

**Comment by nethershaw** [comment]
Don't @ me. You respond to the person you ignored.
